src/s2geography/geoarrow.cc

```diff
@@ -672,6 +672,364 @@ void Reader::ReadGeography(const ArrowArray* array, int64_t offset,
+      S2LatLng ll(loop->vertex(i));
+      coords_[0] = ll.lng().degrees();
+      coords_[1] = ll.lat().degrees();
+      GEOARROW_RETURN_NOT_OK(visitor_.coords(&visitor_, &coords_view_));
+    }
+
+    S2LatLng ll(loop->vertex(loop->num_vertices() - 1));
+    coords_[0] = ll.lng().degrees();
+    coords_[1] = ll.lat().degrees();
+    GEOARROW_RETURN_NOT_OK(visitor_.coords(&visitor_, &coords_view_));
+
+    GEOARROW_RETURN_NOT_OK(visitor_.ring_end(&visitor_));
+    return GEOARROW_OK;
+  }
+
+  int VisitPolygonShell(const S2Polygon& poly, int loop_start) {
+    const S2Loop* loop0 = poly.loop(loop_start);
+    GEOARROW_RETURN_NOT_OK(VisitLoopShell(loop0));
+    for (int j = loop_start + 1; j <= poly.GetLastDescendant(loop_start); j++) {
+      const S2Loop* loop = poly.loop(j);
+      if (loop->depth() == (loop0->depth() + 1)) {
+        GEOARROW_RETURN_NOT_OK(VisitLoopHole(loop));
+      }
+    }
+
+    return GEOARROW_OK;
+  }
+
+  int VisitPolygons(const PolygonGeography& geog) {
+    const S2Polygon& poly = *geog.Polygon();
+
+    // find the outer shells (loop depth = 0, 2, 4, etc.)
+    std::vector<int> outer_shell_loop_ids;
+
+    outer_shell_loop_ids.reserve(poly.num_loops());
+    for (int i = 0; i < poly.num_loops(); i++) {
+      if ((poly.loop(i)->depth() % 2) == 0) {
+        outer_shell_loop_ids.push_back(i);
+      }
+    }
+
+    if (outer_shell_loop_ids.size() == 0) {
+      // empty Polygon
+      GEOARROW_RETURN_NOT_OK(visitor_.geom_start(
+          &visitor_, GEOARROW_GEOMETRY_TYPE_POLYGON, GEOARROW_DIMENSIONS_XY));
+      GEOARROW_RETURN_NOT_OK(visitor_.geom_end(&visitor_));
+
+    } else if (outer_shell_loop_ids.size() == 1) {
+      // Polygon
+      GEOARROW_RETURN_NOT_OK(visitor_.geom_start(
+          &visitor_, GEOARROW_GEOMETRY_TYPE_POLYGON, GEOARROW_DIMENSIONS_XY));
+      GEOARROW_RETURN_NOT_OK(VisitPolygonShell(poly, outer_shell_loop_ids[0]));
+      GEOARROW_RETURN_NOT_OK(visitor_.geom_end(&visitor_));
+
+    } else {
+      // MultiPolygon
+      GEOARROW_RETURN_NOT_OK(
+          visitor_.geom_start(&visitor_, GEOARROW_GEOMETRY_TYPE_MULTIPOLYGON,
+                              GEOARROW_DIMENSIONS_XY));
+      for (size_t i = 0; i < outer_shell_loop_ids.size(); i++) {
+        GEOARROW_RETURN_NOT_OK(visitor_.geom_start(
+            &visitor_, GEOARROW_GEOMETRY_TYPE_POLYGON, GEOARROW_DIMENSIONS_XY));
+        GEOARROW_RETURN_NOT_OK(
+            VisitPolygonShell(poly, outer_shell_loop_ids[i]));
+        GEOARROW_RETURN_NOT_OK(visitor_.geom_end(&visitor_));
+      }
+      GEOARROW_RETURN_NOT_OK(visitor_.geom_end(&visitor_));
+    }
+
+    return GEOARROW_OK;
+  }
+
+  int VisitCollection(const GeographyCollection& geog) {
+    GEOARROW_RETURN_NOT_OK(visitor_.geom_start(
+        &visitor_, GEOARROW_GEOMETRY_TYPE_GEOMETRYCOLLECTION,
+        GEOARROW_DIMENSIONS_XY));
+
+    for (const auto& child_geog : geog.Features()) {
+      auto child_point = dynamic_cast<const PointGeography*>(child_geog.get());
```

Comment:
For spherely, we recently added the type as essentially an attribute on the spherely::Geography C++ class (i.e. which otherwise wraps the s2geography::Geography class): https://github.com/benbovy/spherely/blob/4e39dee1bf46c9591c68bd2b64541ff7c2f8b14f/src/geography.cpp#L116-L162, so whenever a Geography object gets constructed we also store the type (https://github.com/benbovy/spherely/blob/4e39dee1bf46c9591c68bd2b64541ff7c2f8b14f/src/geography.hpp#L55-L59). Not sure if something similar could/should be moved upstream at some point. 

(the clone method a bit higher up in the file is certainly a candidate of moving into s2geography)